design/design-doc-composition-functions.md

```diff
@@ -175,32 +179,557 @@ metadata:
+think of this as a special kind of Composition Function SDK that builds a
+Composition Function from an OCI container, rather than building a Function a
+literal _function_ (e.g. written in Go).
+
+![Crossplane calling two containerized Functions via gRPC](assets/design-doc-composition-functions/containerized-functions.png)
+
+Under this design each containerized Function is "wrapped" in an `xfn`-like
+Composition Function. This means each Function exists as its own Deployment.
+
+This design merges the Protobuf `RunFunctionRequest` with the YAML `FunctionIO`
+types from the alpha design, so instead of reading a `FunctionIO` from stdin a
+containerized Function would read a JSON-encoded `RunFunctionRequest`, and write
+a `RunFunctionResponse` to stdout.
+
+Building a containerized Function would work as follows.
+
+Start with a simple container that reads a JSON `RunFunctionRequest` on stdin
+and writes a JSON `RunFunctionResponse` to stdout. This one simply adds a label
+(credit to @pedjak in https://github.com/crossplane/crossplane/issues/4293) for
+the idea).
+
+```Dockerfile
+FROM alpine:3.18
+RUN apk add --no-cache jq
+ENTRYPOINT ["/bin/sh", "-c", "jq '(.desired.resources[] | .resource.metadata.labels) |= {\"labelizer.xfn.crossplane.io/processed\": \"true\"} + .' | .desired"]
+```
+
+Then build a Function from the image. We want the image to run in its own
+container, not the Function container. This allows the OCI container to run
+within its own namespaces without interfering with (or being interfered with by)
+the `xfn`-like 'adaptor'. To do this we bake the container into the Function
+artifact:
+
+```shell
+# Build your OCI image.
+$ docker build .
+Successfully built 202dc6e5df4c
+
+# Save your OCI image to a tarball.
+$ docker save | gzip > function.tgz
+
+# Add some package metadata for your Function.
+$ vim crossplane.yaml
+
+# Build a Function from the tarball.
+$ kubectl crossplane function build --from-oci-tarball=function.tgz
+
+# Push the package to any OCI registry.
+$ kubectl crossplane function push xpkg.upbound.io/negz/function-labelizer:v0.1.0
+```
+
+The `function build --from-oci-tarball` command creates a Function powered by an
+`xfn`-like adaptor that:
+
+1. Prepares an OCI runtime bundle from the supplied tarball at startup.
+2. Listens for `RunFunctionRequest` RPCs.
+3. Prepares `RunFunctionResponse` RPCs by using an OCI runtime (e.g. `crun`) to
+   invoke the embedded Function in a rootless container.
+
+Note that most of this functionality already exists in `xfn`. This makes it
+cheap enough to implement this Function variant that I believe it's worth doing,
+even if it ends up covering a relatively niche case.
+
+### Using a Function
+
+Composition Functions are very flexible, so the user experience could vary a
+lot. However the general pattern is:
+
+1. Find (or build!) a Function you want to use.
+2. Install the Function using the package manager.
```

Comment:
Can functions be a dependency of a Configuration package? 🤔 - so users won't have to install the Function explicitly, it will be installed automatically through dependency fulfillment when their Configuration package is installed?

Reply:
Yes absolutely. I imagine the dependency relationship might become Configuration -> Function -> Provider in a lot of cases (as opposed to Configuration -> Provider today). I'll try find somewhere to work that in.